src/processor/operator/recursive_extend/bfs_scheduler.cpp

```diff
@@ -0,0 +1,192 @@
+#include "processor/operator/recursive_extend/bfs_scheduler.h"
+
+#include "chrono"
+
+namespace kuzu {
+namespace processor {
+
+/**
+ * Main function of the scheduler that dispatches morsels from the FTable between threads. Or it can
+ * be called if a thread could not find a morsel from it's local SSSPSharedState.
+ *
+ * @return - TRUE indicates check the combination value of the state pair, FALSE indicates some
+ * work was found hence the state's values can be ignored.
+ */
+bool MorselDispatcher::getBFSMorsel(
+    const std::shared_ptr<FTableSharedState>& inputFTableSharedState,
+    std::vector<common::ValueVector*> vectorsToScan, std::vector<ft_col_idx_t> colIndicesToScan,
+    common::ValueVector* srcNodeIDVector, std::unique_ptr<BaseBFSMorsel>& bfsMorsel,
+    std::pair<GlobalSSSPState, SSSPLocalState>& state) {
+    std::unique_lock lck{mutex};
+    switch (schedulerType) {
+    case SchedulerType::OneThreadOneMorsel: {
+        auto inputFTableMorsel = inputFTableSharedState->getMorsel();
+        if (inputFTableMorsel->numTuples == 0) {
+            state = {COMPLETE, NO_WORK_TO_SHARE};
+            return true;
+        }
+        inputFTableSharedState->getTable()->scan(vectorsToScan, inputFTableMorsel->startTupleIdx,
+            inputFTableMorsel->numTuples, colIndicesToScan);
+        bfsMorsel->resetState();
+        auto nodeID = srcNodeIDVector->getValue<common::nodeID_t>(
+            srcNodeIDVector->state->selVector->selectedPositions[0]);
+        bfsMorsel->markSrc(nodeID);
+        return false;
+    }
+    case SchedulerType::nThreadkMorsel: {
+        switch (globalState) {
+        case COMPLETE: {
+            state = {globalState, NO_WORK_TO_SHARE};
+            return true;
+        }
+        case IN_PROGRESS_ALL_SRC_SCANNED: {
+            return findAvailableSSSP(bfsMorsel, state);
+        }
+        case IN_PROGRESS: {
+            if (numActiveSSSP < activeSSSPSharedState.size()) {
+                auto inputFTableMorsel = inputFTableSharedState->getMorsel();
+                if (inputFTableMorsel->numTuples == 0) {
+                    globalState = (numActiveSSSP == 0) ? COMPLETE : IN_PROGRESS_ALL_SRC_SCANNED;
+                    state = {globalState, NO_WORK_TO_SHARE};
+                    return true;
+                }
+                numActiveSSSP++;
+                auto newSSSPSharedState =
```

Comment:
I would avoid creating new object and instead always reuse the same SSSPSharedState object.

Think in this way. The maximum SSSPSharedState you can have at the same time equals to the maximum number of threads. You should create the object at compile time. At runtime only write to it. This is database processing principal, avoid creating new object as much as you can.

Reply:
>always reuse the same SSSPSharedState object

Yeah this has 1 major problem, if you have 4 threads working on the same SSSP and then 3 of them finished their work and try to reset the existing SSSP object, the other straggler thread will not be aware of this.
It might still be doing some extension work and end up updating the `visited` vector of this new SSSP.

Something similar could happen let's say for path Length writing step where 3 of them finished writing the lengths to their ValueVectors and then reset the destination vector while the 4th thread was still reading it leading to some values getting lost.

One way I can think of to handle this is to not only check if a morsel is `MORSEL_COMPLETE` but also if `numThreadsActiveOnThread` is 0 or not. This should handle the 1st case. For the 2nd case, there is no thread counter currently for tracking who is helping.

Reply:
As per our discussion, I've changed this part's logic to reuse existing SharedState objects instead of creating a new one everytime. And for thread safety 3 conditions are being checked, before deciding which position to reuse.